hibernate-core/src/main/java/org/hibernate/internal/util/collections/PagedArray.java

```diff
@@ -0,0 +1,255 @@
+/*
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.internal.util.collections;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+/**
+ * Utility implementation of an array-like structure organized in sub-arrays (pages)
+ * that can grow efficiently by a constant factor of {@link #PAGE_CAPACITY}.
+ * Reading from and writing to the array is a simple O(2) operation.
+ *
+ * @param <E> the type of elements stored in this array
+ * @implNote To optimize the memory footprint of the array, removed indexes should be reused
+ * to have the least number of pages allocated as possible.
+ */
+public class PagedArray<E> {
+	// It's important that capacity is a power of 2 to allow calculating page index and offset within the page
+	// with simple division and modulo operations; also static final so JIT can inline these operations.
+	private static final int PAGE_CAPACITY = 1 << 5; // 32
+	private static final int PAGE_SHIFT = Integer.numberOfTrailingZeros( PAGE_CAPACITY );
```

Comment:
shift and masks are not needed if PAGE_CAPACITY is `static final` and a power of 2; the JIT will take care of replacing the division and modulus with optimized ops (it's called "strenght reduction" optimization)

Reply:
Oh great, I thought you still had to cache those! I'll replace them with simple division and modulo operations than, thanks!